[OAP-BanyanDB][Bug] Error on Creating Schema Concurrently

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.


### Apache SkyWalking Component

OAP server (apache/skywalking)

### What happened

```
 Caused by: org.apache.skywalking.banyandb.v1.client.grpc.exception.AlreadyExistsException: io.grpc.StatusRuntimeException: ALREADY_EXISTS: banyandb: resource already exists                            │
│     at org.apache.skywalking.banyandb.v1.client.grpc.exception.BanyanDBApiExceptionFactory.createException(BanyanDBApiExceptionFactory.java:38) ~[banyandb-java-client-0.6.0.jar:0.6.0]                 │
│     at org.apache.skywalking.banyandb.v1.client.grpc.exception.BanyanDBGrpcApiExceptionFactory.create(BanyanDBGrpcApiExceptionFactory.java:53) ~[banyandb-java-client-0.6.0.jar:0.6.0]                  │
│     at org.apache.skywalking.banyandb.v1.client.grpc.exception.BanyanDBGrpcApiExceptionFactory.createException(BanyanDBGrpcApiExceptionFactory.java:41) ~[banyandb-java-client-0.6.0.jar:0.6.0]         │
│     at org.apache.skywalking.banyandb.v1.client.grpc.HandleExceptionsWith.callAndTranslateApiException(HandleExceptionsWith.java:47) ~[banyandb-java-client-0.6.0.jar:0.6.0]                            │
│     at org.apache.skywalking.banyandb.v1.client.grpc.MetadataClient.execute(MetadataClient.java:100) ~[banyandb-java-client-0.6.0.jar:0.6.0]                                                            │
│     at org.apache.skywalking.banyandb.v1.client.metadata.GroupMetadataRegistry.create(GroupMetadataRegistry.java:40) ~[banyandb-java-client-0.6.0.jar:0.6.0]                                            │
│     at org.apache.skywalking.banyandb.v1.client.BanyanDBClient.define(BanyanDBClient.java:410) ~[banyandb-java-client-0.6.0.jar:0.6.0]                                                                  │
│     at org.apache.skywalking.oap.server.storage.plugin.banyandb.MetadataRegistry$SchemaMetadata.checkResourceExistence(MetadataRegistry.java:554) ~[storage-banyandb-plugin-10.1.0-SNAPSHOT.jar:10.1.0- │
│     at org.apache.skywalking.oap.server.storage.plugin.banyandb.BanyanDBIndexInstaller.isExists(BanyanDBIndexInstaller.java:56) ~[storage-banyandb-plugin-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]          │
│     at org.apache.skywalking.oap.server.core.storage.model.ModelInstaller.whenCreating(ModelInstaller.java:53) ~[server-core-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]                                       │
│     at org.apache.skywalking.oap.server.core.storage.model.StorageModels.addModelListener(StorageModels.java:141) ~[server-core-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]                                    │
│     at org.apache.skywalking.oap.server.storage.plugin.banyandb.BanyanDBStorageProvider.start(BanyanDBStorageProvider.java:181) ~[storage-banyandb-plugin-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]          │
│     ... 4 more                                                                                                                                                                                          │
│ Caused by: io.grpc.StatusRuntimeException: ALREADY_EXISTS: banyandb: resource already exists                                                                                                            │
│     at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:268) ~[grpc-stub-1.63.0.jar:1.63.0]                                                                                           │
│     at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:249) ~[grpc-stub-1.63.0.jar:1.63.0]                                                                                                       │
│     at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:167) ~[grpc-stub-1.63.0.jar:1.63.0]                                                                                                  │
│     at org.apache.skywalking.banyandb.database.v1.GroupRegistryServiceGrpc$GroupRegistryServiceBlockingStub.create(GroupRegistryServiceGrpc.java:396) ~[banyandb-java-client-0.6.0.jar:0.6.0]           │
│     at org.apache.skywalking.banyandb.v1.client.metadata.GroupMetadataRegistry.lambda$create$0(GroupMetadataRegistry.java:40) ~[banyandb-java-client-0.6.0.jar:0.6.0]                                   │
│     at org.apache.skywalking.banyandb.v1.client.grpc.HandleExceptionsWith.callAndTranslateApiException(HandleExceptionsWith.java:45) ~[banyandb-java-client-0.6.0.jar:0.6.0]                            │
│     at org.apache.skywalking.banyandb.v1.client.grpc.MetadataClient.execute(MetadataClient.java:100) ~[banyandb-java-client-0.6.0.jar:0.6.0]                                                            │
│     at org.apache.skywalking.banyandb.v1.client.metadata.GroupMetadataRegistry.create(GroupMetadataRegistry.java:40) ~[banyandb-java-client-0.6.0.jar:0.6.0]                                            │
│     at org.apache.skywalking.banyandb.v1.client.BanyanDBClient.define(BanyanDBClient.java:410) ~[banyandb-java-client-0.6.0.jar:0.6.0]                                                                  │
│     at org.apache.skywalking.oap.server.storage.plugin.banyandb.MetadataRegistry$SchemaMetadata.checkResourceExistence(MetadataRegistry.java:554) ~[storage-banyandb-plugin-10.1.0-SNAPSHOT.jar:10.1.0- │
│     at org.apache.skywalking.oap.server.storage.plugin.banyandb.BanyanDBIndexInstaller.isExists(BanyanDBIndexInstaller.java:56) ~[storage-banyandb-plugin-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]          │
│     at org.apache.skywalking.oap.server.core.storage.model.ModelInstaller.whenCreating(ModelInstaller.java:53) ~[server-core-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]                                       │
│     at org.apache.skywalking.oap.server.core.storage.model.StorageModels.addModelListener(StorageModels.java:141) ~[server-core-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]                                    │
│     at org.apache.skywalking.oap.server.storage.plugin.banyandb.BanyanDBStorageProvider.start(BanyanDBStorageProvider.java:181) ~[storage-banyandb-plugin-10.1.0-SNAPSHOT.jar:10.1.0-SNAPSHOT]  
```

### What you expected to happen

No error

### How to reproduce

Run no more than 1 OAP to concurrently create the schema.

### Anything else

_No response_

### Are you willing to submit a pull request to fix on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/docs/en/changes/changes.md
+++ b/docs/en/changes/changes.md
@@ -26,6 +26,7 @@
 * Support `labelCount` function in the OAL engine.
 * Support BanyanDB internal metrics query execution tracing.
 * BanyanDB client config: rise the default `maxBulkSize` to 10000, add `flushTimeout` and set default to 10s.
+* Polish BanyanDB group and schema creation logic to fix the schema creation failure issue in distributed race conditions.
 
 #### UI
 * Highlight search log keywords.

--- a/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBIndexInstaller.java
+++ b/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBIndexInstaller.java
@@ -18,6 +18,7 @@
 
 package org.apache.skywalking.oap.server.storage.plugin.banyandb;
 
+import io.grpc.Status;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.skywalking.banyandb.v1.client.BanyanDBClient;
 import org.apache.skywalking.banyandb.v1.client.grpc.exception.BanyanDBException;
@@ -48,8 +49,11 @@ public class BanyanDBIndexInstaller extends ModelInstaller {
         if (!model.isTimeSeries()) {
             return true;
         }
-        final ConfigService configService = moduleManager.find(CoreModule.NAME).provider().getService(ConfigService.class);
-        final MetadataRegistry.SchemaMetadata metadata = MetadataRegistry.INSTANCE.parseMetadata(model, config, configService);
+        final ConfigService configService = moduleManager.find(CoreModule.NAME)
+                                                         .provider()
+                                                         .getService(ConfigService.class);
+        final MetadataRegistry.SchemaMetadata metadata = MetadataRegistry.INSTANCE.parseMetadata(
+            model, config, configService);
         try {
             final BanyanDBClient c = ((BanyanDBStorageClient) this.client).client;
             // first check resource existence and create group if necessary
@@ -78,26 +82,60 @@ public class BanyanDBIndexInstaller extends ModelInstaller {
     @Override
     public void createTable(Model model) throws StorageException {
         try {
-            ConfigService configService = moduleManager.find(CoreModule.NAME).provider().getService(ConfigService.class);
+            ConfigService configService = moduleManager.find(CoreModule.NAME)
+                                                       .provider()
+                                                       .getService(ConfigService.class);
             if (model.isRecord()) { // stream
                 Stream stream = MetadataRegistry.INSTANCE.registerStreamModel(model, config, configService);
                 if (stream != null) {
                     log.info("install stream schema {}", model.getName());
-                    ((BanyanDBStorageClient) client).define(stream);
+                    try {
+                        ((BanyanDBStorageClient) client).define(stream);
+                    } catch (BanyanDBException ex) {
+                        if (ex.getStatus().equals(Status.Code.ALREADY_EXISTS)) {
+                            log.info(
+                                "Stream schema {}_{} already created by another OAP node",
+                                model.getName(),
+                                model.getDownsampling()
+                            );
+                        } else {
+                            throw ex;
+                        }
+                    }
                 }
             } else { // measure
                 Measure measure = MetadataRegistry.INSTANCE.registerMeasureModel(model, config, configService);
                 if (measure != null) {
                     log.info("install measure schema {}", measure.name());
-                    ((BanyanDBStorageClient) client).define(measure);
                     final BanyanDBClient c = ((BanyanDBStorageClient) this.client).client;
-                    MetadataRegistry.INSTANCE.findMetadata(model).installTopNAggregation(c);
+                    try {
+                        c.define(measure);
+                    } catch (BanyanDBException ex) {
+                        if (ex.getStatus().equals(Status.Code.ALREADY_EXISTS)) {
+                            log.info("Measure schema {}_{} already created by another OAP node",
+                                     model.getName(),
+                                     model.getDownsampling());
+                        } else {
+                            throw ex;
+                        }
+                    }
+                    final MetadataRegistry.Schema schema = MetadataRegistry.INSTANCE.findMetadata(model);
+                    try {
+                        schema.installTopNAggregation(c);
+                    } catch (BanyanDBException ex) {
+                        if (ex.getStatus().equals(Status.Code.ALREADY_EXISTS)) {
+                            log.info("Measure schema {}_{} TopN({}) already created by another OAP node",
+                                     model.getName(),
+                                     model.getDownsampling(),
+                                     schema.getTopNSpec());
+                        } else {
+                            throw ex;
+                        }
+                    }
                 }
             }
-        } catch (IOException ex) {
-            throw new StorageException("fail to install schema", ex);
         } catch (BanyanDBException ex) {
-            throw new StorageException("fail to install TopN schema", ex);
+            throw new StorageException("fail to create schema " + model.getName(), ex);
         }
     }
 }

--- a/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBIndexInstaller.java
+++ b/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBIndexInstaller.java
@@ -32,8 +32,6 @@ import org.apache.skywalking.oap.server.core.storage.model.ModelInstaller;
 import org.apache.skywalking.oap.server.library.client.Client;
 import org.apache.skywalking.oap.server.library.module.ModuleManager;
 
-import java.io.IOException;
-
 @Slf4j
 public class BanyanDBIndexInstaller extends ModelInstaller {
     private final BanyanDBStorageConfig config;

--- a/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBStorageClient.java
+++ b/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBStorageClient.java
@@ -180,23 +180,23 @@ public class BanyanDBStorageClient implements Client, HealthCheckable {
         }
     }
 
-    public void define(Stream stream) throws IOException {
+    public void define(Stream stream) throws BanyanDBException {
         try {
             this.client.define(stream);
             this.healthChecker.health();
         } catch (BanyanDBException ex) {
             healthChecker.unHealth(ex);
-            throw new IOException("fail to define stream", ex);
+            throw ex;
         }
     }
 
-    public void define(Measure measure) throws IOException {
+    public void define(Measure measure) throws BanyanDBException {
         try {
             this.client.define(measure);
             this.healthChecker.health();
         } catch (BanyanDBException ex) {
             healthChecker.unHealth(ex);
-            throw new IOException("fail to define stream", ex);
+            throw ex;
         }
     }
 

--- a/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/MetadataRegistry.java
+++ b/oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/MetadataRegistry.java
@@ -551,22 +551,44 @@ public enum MetadataRegistry {
                 case STREAM:
                     resourceExist = client.existStream(this.group, this.name());
                     if (!resourceExist.hasGroup()) {
-                        Group g = client.define(Group.create(this.group, Catalog.STREAM, this.shard,
-                                IntervalRule.create(IntervalRule.Unit.DAY, this.segmentIntervalDays),
-                                IntervalRule.create(IntervalRule.Unit.DAY, this.ttlDays)));
-                        if (g != null) {
-                            log.info("group {} created", g.name());
+                        try {
+                            Group g = client.define(Group.create(this.group, Catalog.STREAM, this.shard,
+                                                                 IntervalRule.create(
+                                                                     IntervalRule.Unit.DAY, this.segmentIntervalDays),
+                                                                 IntervalRule.create(
+                                                                     IntervalRule.Unit.DAY, this.ttlDays)
+                            ));
+                            if (g != null) {
+                                log.info("group {} created", g.name());
+                            }
+                        } catch (BanyanDBException ex) {
+                            if (ex.getStatus().equals(Status.Code.ALREADY_EXISTS)) {
+                                log.info("group {} already created by another OAP node", this.group);
+                            } else {
+                                throw ex;
+                            }
                         }
                     }
                     return resourceExist.hasResource();
                 case MEASURE:
                     resourceExist = client.existMeasure(this.group, this.name());
-                    if (!resourceExist.hasGroup()) {
-                        Group g = client.define(Group.create(this.group, Catalog.MEASURE, this.shard,
-                                IntervalRule.create(IntervalRule.Unit.DAY, this.segmentIntervalDays),
-                                IntervalRule.create(IntervalRule.Unit.DAY, this.ttlDays)));
-                        if (g != null) {
-                            log.info("group {} created", g.name());
+                    try {
+                        if (!resourceExist.hasGroup()) {
+                            Group g = client.define(Group.create(this.group, Catalog.MEASURE, this.shard,
+                                                                 IntervalRule.create(
+                                                                     IntervalRule.Unit.DAY, this.segmentIntervalDays),
+                                                                 IntervalRule.create(
+                                                                     IntervalRule.Unit.DAY, this.ttlDays)
+                            ));
+                            if (g != null) {
+                                log.info("group {} created", g.name());
+                            }
+                        }
+                    } catch (BanyanDBException ex) {
+                        if (ex.getStatus().equals(Status.Code.ALREADY_EXISTS)) {
+                            log.info("group {} already created by another OAP node", this.group);
+                        } else {
+                            throw ex;
                         }
                     }
                     return resourceExist.hasResource();

